File: src/pieces/QueenBehaviour.rs
use crate::{types::{Array2D, Position, Shape, PositionContent}, Array::{sort_by_distance, get_diagonal_positions, get_col_row_positions}};

use super::piece::Piece;

pub(crate) fn allowed_moves(
    piece: &Piece,
    board: &Array2D,
    pos: &Position,
    shape: &Shape,
    is_white: bool,
) -> Vec<Position> {
    let col_row_positions = get_col_row_positions(pos, shape);
    let diagonal_positions = get_diagonal_positions(pos, shape);
    let mut position_lines = Vec::with_capacity(2 + diagonal_positions.len());
    for positions in col_row_positions.into_iter() {
        position_lines.push(positions)
    }
    for positions in diagonal_positions.into_iter() {
        position_lines.push(positions)
    }
    let mut filt_all_ms: Vec<Position> = Vec::with_capacity(20);
    for positions in position_lines.into_iter() {
        for dir_positions in positions {
            let dir_positions = sort_by_distance(pos, dir_positions);
            for move_pos in dir_positions.into_iter() {
                let mp_piece: &PositionContent = &board[usize::try_from(move_pos.0).unwrap()]
                    [usize::try_from(move_pos.1).unwrap()];
                match mp_piece {
                    PositionContent::PieceContent(_) => filt_all_ms.push(move_pos),
                    PositionContent::Empty => break,
                }
            }
        }
    }
    filt_all_ms
}

pub(crate) fn allowed_takes(
    piece: &Piece,
    board: &Array2D,
    pos: &Position,
    shape: &Shape,
    is_white: bool,
) -> Vec<Position> {
    let col_row_positions = get_col_row_positions(pos, shape);
    let diagonal_positions = get_diagonal_positions(pos, shape);
    let mut position_lines = Vec::with_capacity(2 + diagonal_positions.len());
    for positions in col_row_positions.into_iter() {
        position_lines.push(positions)
    }
    for positions in diagonal_positions.into_iter() {
        position_lines.push(positions)
    }
    let mut filt_all_ts: Vec<Position> = Vec::with_capacity(20);
    for positions in position_lines.into_iter() {
        for dir_positions in positions {
            let dir_positions = sort_by_distance(pos, dir_positions);
            for take_pos in dir_positions.into_iter() {
                let tp_piece: &PositionContent = &board[usize::try_from(take_pos.0).unwrap()]
                    [usize::try_from(take_pos.1).unwrap()];
                match tp_piece {
                    PositionContent::PieceContent(tp_piece) => {
                        if tp_piece.is_white != piece.is_white {
                            filt_all_ts.push(take_pos)
                        }
                        break
                    },
                    PositionContent::Empty => break,
                }
            }
        }
    }
    filt_all_ts
}

File: src/pieces/QueenBehaviour.ts

function queen_allowed_moves(piece: Piece, board: Array2D, pos: Position, is_white: boolean): Position[] {
    const col_row_positions = get_col_row_positions(pos, board.shape)
    const diagonal_positions = get_diagonal_positions(pos, board.shape)
    const position_lines = [col_row_positions, diagonal_positions].flat()
    let filt_all_ms: Position[] = []
    for (const positions of position_lines) {
        for (let dir_positions of positions) {
            dir_positions.sort()
            for (const move_pos of dir_positions) {
                const mp_piece = board.data[move_pos[0]][move_pos[1]]
                if (mp_piece.is_empty) {
                    filt_all_ms.push(move_pos)
                } else {
                    break
                }
            }
        }
    }
    return filt_all_ms
}

function queen_allowed_takes(piece: Piece, board: Array2D, pos: Position, is_white: boolean): Position[] {
    const col_row_positions = get_col_row_positions(pos, board.shape)
    const diagonal_positions = get_diagonal_positions(pos, board.shape)
    const position_lines = [col_row_positions, diagonal_positions].flat()
    let filt_all_ms: Position[] = []
    for (const positions of position_lines) {
        for (let dir_positions of positions) {
            dir_positions.sort()
            for (const move_pos of dir_positions) {
                const tp_piece = board.data[move_pos[0]][move_pos[1]]
                if (!tp_piece.is_empty) {
                    if (tp_piece.is_white !== piece.is_white) {
                        filt_all_ms.push(move_pos)
                    }
                    break
                }
            }
        }
    }
    return filt_all_ms}